### PR TITLE
cache structure for RRSet with ranking

### PR DESCRIPTION
**Pull request by khibino**
add data structure and interfaces to cache RRSet with ranking

- [x] cache RRSet with consistent TTL.
- [x] test properties for cache.
- [x] ranking implementation. get section RRs with rank.
- [x] test for ranking.

### CODE REVIEW

File: src/DNSC/Cache.hs [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+  where
+    result k ttl (V crs rank) = (extractRRSet k ttl crs, rank)
+
+lookup_ :: Timestamp -> (Key -> TTL -> Val -> a)
+        -> CDomain -> TYPE -> CLASS
+        -> Cache -> Maybe a
+lookup_ now mk dom typ cls (Cache lifetimes crss) = do
+  let k = K dom typ cls
+  eol <- k `PSQ.lookup` lifetimes
+  ttl <- alive now eol
```
Comment (khibino) [inline src/DNSC/Cache.hs:158]: 経過時間分、TTL が減少する

File: src/DNSC/Cache.hs [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+import Data.IP (IPv4, IPv6)
+import Network.DNS
+  (Domain, CLASS, TTL, TYPE (..), RData (..),
+   ResourceRecord (ResourceRecord), DNSMessage)
+import qualified Network.DNS as DNS
+
+
+---
+
+type CDomain = ShortByteString
```
Comment (kazu-yamamoto) [inline src/DNSC/Cache.hs:54]: Very good to avoid fragmentation.

File: src/DNSC/Cache.hs [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+  | CR_PTR [CDomain]
+  | CR_MX [(Word16, CDomain)]
+  | CR_TXT [CTxt]
+  | CR_AAAA [IPv6]
+  deriving (Eq, Ord, Show)
+
+---
+
+-- Ranking data (section 5.4.1 of RFC2181 - Clarifications to the DNS Specification)
+-- <https://datatracker.ietf.org/doc/html/rfc2181#section-5.4.1>
```
Comment (kazu-yamamoto) [inline src/DNSC/Cache.hs:73]: Since DNS-related RFCs are messy, this kind of comment is very useful.

File: src/DNSC/Cache.hs [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+rankedAdditional =
+  rankedSection
+  (Just rankAdditional)
+  (Just rankAdditional)
+  DNS.additional
+
+---
+
+data Key = K CDomain TYPE CLASS deriving (Eq, Ord, Show)
+data Val = V CRSet Ranking deriving Show
```
Comment (kazu-yamamoto) [inline src/DNSC/Cache.hs:136]: `K` and `V` are exported. I guess you'd better to choose more meaningful names.

Comment (khibino) [reply src/DNSC/Cache.hs]: https://github.com/khibino/dns-resolver/pull/2/commits/c93432cfe42f7fd2ba61d7c99d0d831eb4757782
https://github.com/khibino/dns-resolver/pull/2/commits/afc58b5a7b3fca74489f8927fb86263f419a0584

**Review by kazu-yamamoto — APPROVED** [review]
LGTM

**Pull request merged**